numba/np/ufunc/gufunc.py

```diff
@@ -0,0 +1,130 @@
+        self.ufunc = self.gufunc_builder.build_ufunc()
+        return self
+
+    def disable_compile(self):
+        assert len(self.gufunc_builder._sigs) > 0
+        self._frozen = True
+
+    @property
+    def __doc__(self):
+        return self.ufunc.__doc__
+
+    @property
+    def __name__(self):
+        return self.ufunc.__name__
+
+    @property
+    def nin(self):
+        return self.ufunc.nin
+
+    @property
+    def nout(self):
+        return self.ufunc.nout
+
+    @property
+    def nargs(self):
+        return self.ufunc.nargs
+
+    @property
+    def ntypes(self):
+        return self.ufunc.ntypes
+
+    @property
+    def types(self):
+        return self.ufunc.types
+
+    @property
+    def identity(self):
+        return self.ufunc.identity
+
+    def _get_ewise_dtypes(self, args):
+        argtys = map(lambda x: typeof(x), args)
+        tys = []
+        for argty in argtys:
+            if isinstance(argty, types.Array):
+                tys.append(argty.dtype)
+            else:
+                tys.append(argty)
+        return tys
+
+    def _get_signature(self, *args):
+        parsed_sig = parse_signature(self.gufunc_builder.signature)
+        ewise_types = self._get_ewise_dtypes(args)  # [int32, int32, int32, ...]  # noqa: E501
+
+        # Two cases here can happen here
+        # 1. args contains the output array:
+        #      gufunc(A, B, C)
+        # 2. args doesn't contains the output array:
+        #      C = gufunc(A, B)
+        #
+        # In the latter case, one would have to guess the type
+        # and format (scalar or array). This behavior will be
+        # forbidden for now.
+        # See: https://github.com/numba/numba/pull/5938#issuecomment-661978921
+
+        # parsed_sig[1] has always length 1
+        if len(ewise_types) < len(parsed_sig[0]) + 1:
+            msg = (
+                f"Too few arguments for function '{self.__name__}'. "
+                "Note that the pattern `out = gufunc(Arg1, Arg2, ..., ArgN)` "
+                "is not allowed. Use `gufunc(Arg1, Arg2, ..., ArgN, out) instead.")  # noqa: E501
+            raise TypeError(msg)
+
+        # first time calling the gufunc
+        # generate a signature based on input arguments
+        l = []
+        for idx, sig_dim in enumerate(parsed_sig[0]):
+            ndim = len(sig_dim)
+            if ndim == 0:  # append scalar
+                l.append(ewise_types[idx])
+            else:
+                l.append(types.Array(ewise_types[idx], ndim, 'A'))
+
+        # add return type to signature
+        retty = ewise_types[-1]
+        ret_ndim = len(parsed_sig[-1][0]) or 1  # small hack to return scalar
+        l.append(types.Array(retty, ret_ndim, 'A'))
+
+        return types.none(*l)
+
+    def __call__(self, *args):
+
+        if self._frozen:
+            return self.ufunc(*args)
+
+        ewise = self._get_ewise_dtypes(args)
+        if not (self.ufunc and ufunc_find_matching_loop(self.ufunc, ewise)):
```

Comment:
missing check for lack of output argument if this branch is `ufunc_find_matching_loop()` returns True.

Reply:
^ this comment corresponds to https://github.com/numba/numba/pull/5938#discussion_r503534505

Reply:
Hi Siu, If we check for missing lack of output argument when `ufunc_find_matching_loop()` returns True, it will crash tests `test_gufunc.py::TestGUFunc.(test_ufunc_like/test_gufunc)`. This test manually adds signatures by calling `.add` and never calls `disable_compile`.

Reply:
In that case, we need to add the logic to tell apart if it is a dynamic gufunc or not. Otherwise, the behavior will be confusing. 

We need the test in https://github.com/numba/numba/pull/5938/commits/9236b15e1e0a38a1fe23aa3f69d44b359d18abe3 and more tests around the error.

Reply:
Hi @sklam,

Ok, I can do that but do you think this is necessary? For instance, in the example you gave at https://github.com/numba/numba/commit/9236b15e1e0a38a1fe23aa3f69d44b359d18abe3, it is completely ok to omit the `out` argument: `out = sum_row(inp)`. The problem appears only in the first call to it.

Reply:
The error behavior should not depend on the ordering of the calls. It will be confusing to users if their gufunc works sometimes by accident. So, I think it is best to guarantee an error for missing output argument for all dynamic gufuncs.

Reply:
Okay, done! I've moved the typing check part to before any computation is done. This way, numba will check if the number of arguments match in a dynamic gufunc before attempting to compile or call it. Let me know if this approach is what you're looking for.